### ISSUE TITLE
GEP: Client Certificate Verification for Gateway Listeners

### ISSUE THREAD

**Issue by jpeach** [kubernetes-sigs/gateway-api#91]
**What would you like to be added**:

The ability for a HTTPS (or TLS generally) endpoint to require that the client present a certificate that can be validated according to some configurable policy.

**Why is this needed**:

As an application developer, I want to restrict access to my application to a certain audience of clients. The audience is defined by one or more of

- a collection of specific TLS certificates (maybe by hash)
- a collection of subject names in certificates
- a collection of certificates issued by a specific (unique) CA

I want the infrastructure to guarantee that I only receive client traffic that originates from this audience.

/kind user-story

**Comment by bowei** [comment]
Is this mTLS? Can you add to the issue title, as that is a commonly known term around the K8S community.

**Comment by youngnick** [comment]
It is mutual TLS in the sense that the client and the server perform mutual TLS, but 'mTLS' also has some connotations about speed and automation of certificate rotation that we probably don't want to bring in here.

I say this because when we had an issue to add 'mTLS' to Contour, people started asking straight away if Contour was now a service mesh. Which it is not. But the term is overloaded.

**Comment by hbagdi** [comment]
/remove-lifecycle rotten

#268 should close this.

**Comment by alanchrt** [comment]
Adding my comments from Slack regarding some configuration options that would be nice:

> either deny or don't deny requests from clients who don't present valid certs (useful when you want to show a "logged out state" based on application logic in your upstream)

> an option to pass through the subject/sans from the certificate to your upstream (the identity of the client), maybe in a header? not sure if there's a standard there. this could probably be the default.. not sure if it would matter if you still passed through the client identity and the upstream didn't care.

> this could let you configure things like "i'll take requests in my upstream even if they don't present a valid cert, but if they do present a valid cert, i want to know that the identity was alan so i can decide his access level. and if there's no cert, i want to show them logged out state."

The `ingress-nginx` annotations seem like a really good reference point for which configuration options might be nice, as well as the names and contents of headers to pass back to the upstream:

https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#client-certificate-authentication

I think this one largely gets at some of the use cases I was describing above:

```yaml
nginx.ingress.kubernetes.io/auth-tls-verify-client: optional_no_ca
```

**Comment by youngnick** [comment]
It's also important to remember that, if the following are true:
- you have a single proxy routing multiple SNIs
- the multiple SNI routes can have different security profiles (as in, some can have client certificates and some can't)
- You are using a protocol that has a higher-level routing construct (like HTTP's `Host` header)

Then you can connect to `insecure.foo.com` without a client cert, and set your `Host` header to `secure.foo.com`, and end up at the secure site, with no client cert.

This is roughly the same thing as [domain fronting](https://en.wikipedia.org/wiki/Domain_fronting).

**Comment by alanchrt** [comment]
I wonder if sane defaults plus an "I know I'm doing a dangerous thing" option could help with domain fronting.

That is, if your listener is configured in `Terminate` mode, client certificate authentication can be configured with no special options. In addition to verifying the certificate, requests by default won't be routed upstream unless the SNI host matches the HTTP `Host`.

But, if you're in `Passthrough` mode, client authentication is only allowed with `disableHostVerify: true`. The docs can indicate that any upstream routers will need to be sure not to route based on `Host` with the expectation that it is the actually hostname that the client authenticated with. The `disableHostVerify` option could also be available for `Terminate` mode if there was a use case that needed it.

**Comment by mmalone** [comment]
@alanchrt you're talking about two modes: one that enforces that Host matches SNI and the other that doesn't, right? We're not terminating TLS vs. passing it through for a backend to terminate? If that's correct, I like the idea but would probably look for different terminology ;).

I'm not convinced the non-enforcing mode needs to exist at all. What's the use case for upstreams actually _needing_ to route based on Host header? I'd consider _setting_ the Host header for the upstream request to _force_ it to match SNI. The host header is ignored all the time (e.g., most non-proxy web applications will ignore the host header). In a case like this, I can't think of a way it would ever break non-malicious clients.

I don't have as much experience with http/2 so I have no comment on whether this might affect those clients.

**Comment by howardjohn** [comment]
> I don't have as much experience with http/2 so I have no comment on whether this might affect those clients.

I think per the HTTP2 spec servers are expected to return a 421 here in some scenarios - this allows the browser to automatically retry when its optimizations cause it to not match
https://github.com/envoyproxy/envoy/issues/6767 and https://github.com/istio/istio/issues/13589 have a lot of info here

**Comment by alanchrt** [comment]
@mmalone Actually, I wasn't trying to suggest `Terminate` and `Passthrough` as names for client auth handling, but rather trying to describe some possible default behaviors for those modes, which are [defined in the TLS config](https://github.com/kubernetes-sigs/gateway-api/blob/675ca43a6212d902acc8bd055e651816047129b3/apis/v1alpha1/gateway_types.go#L310-L324).

Enforcement of SNI host and HTTP `Host` match wouldn't be possible in a `Passthrough` scenario, which is why I was suggesting you'd have to explicitly disable the enforcement to be able to enable client auth. Then, people know what they're getting themselves into.

As an example of an upstream routing based on `Host`, imagine it wasn't even a router usptream, but rather a monolith web app that's multi-tenant. It could server `customer-a.mycompany.com` and `customer-b.mycompany.com`.

This is totally contrived, but imagine they decided they wanted to have cient cert auth on `Gateway` to prevent Customer A from even being able to load Customer B's app. They already have all the wiring to terminate TLS and manage certs and keys for their upstream app (for the customer domains), so they don't want to mess with that. They just verify the client cert, then passthrough TLS traffic to the app upstream. But, the app uses the HTTP `Host` to determine which customer's app to display.

Again, contrived, but I could see scenarios where someone might want it. In the above scenario, they could set `disableHostVerify: true`, and they could check the SNI host (passed in a header) to make sure it matches the HTTP `Host` in the upstream app itself.

**Comment by shaneutt** [comment]
Despite this issue being quite old, we the maintainers are still pretty convinced that we want to have this functionality in a future release. We are marking this `help wanted` as we're looking for contributors with strong use cases to help champion and drive this forward.

**Comment by evankanderson** [comment]
I think this is related to https://github.com/kubernetes-sigs/gateway-api/discussions/1244 and therefore GEP https://github.com/kubernetes-sigs/gateway-api/issues/1282 might be relevant / close the loop here.

**Comment by evankanderson** [comment]
It's likely this may block future Knative adoption of Gateway API until it's resolved -- we're moving to re-encrypt traffic from the ingress between our components to the user's pod.

**Comment by youngnick** [comment]
It's related in that it's TLS, but this is for client certificate verification on the Listener, so on the connection from the client to the "outside" of the Gateway. #1244 and #1282 are both about the connection from the Gateway to the backend, "inside" the Gateway. (The descriptions are in quotes because they're not always the case, there's more to it than that, etc).

**Comment by arkodg** [comment]
Hoping to help out with this issue.

**Existing Projects**
* [Here's](https://istio.io/latest/docs/tasks/traffic-management/ingress/secure-ingress/#configure-a-mutual-tls-ingress-gateway) an example of how mTLS is described to the user in Istio and how the user can configure their Gateway to support mTLS (by setting the `tls.mode` to `MUTUAL` and specifying the CA Certificate as `cacert` in the same secret where the server key and cert are defined)
* [Here's](https://projectcontour.io/docs/v1.23.0/config/tls-termination/) another example from Contour that asks the user to define a `clientValidation` block where the secret containing the CA Certificate can be defined (the secret must be a  `Opaque` secret with the name `ca.cert`)

**What Common between them**
* Explicitly define client validation along with TLS Termination (`MUTUAL` / `clientValidation`)
* Provide a way to plug in the CA Certificate

**Current State in the Gateway API**
* [Has a mode](https://github.com/kubernetes-sigs/gateway-api/blob/21bba43681b3db3a62d7ca9f108bf3be4c0cea71/apis/v1beta1/gateway_types.go#L305) field to define TLS session type 
* has the ability to define Certificate as secrets of type `tls` with [CertificateRefs](https://github.com/kubernetes-sigs/gateway-api/blob/21bba43681b3db3a62d7ca9f108bf3be4c0cea71/apis/v1beta1/gateway_types.go#L334)

**Follow up questions for the community**
* Should mTLS reuse the `Terminate` mode or should there be a a new mode defined for mTLS
* How does the user plug in the CA Cert
  * In its current state, we cannot reuse `CertificateRefs` because `tls` Secrets dont support a CA Cert field
 * Assumed that the API might not want to expose more knobs such allowing specific SAN Names ([this](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#envoy-v3-api-field-extensions-transport-sockets-tls-v3-certificatevalidationcontext-match-typed-subject-alt-names) is what Envoy exposes), please correct me if that is not the case

**Comment by arkodg** [comment]
@shaneutt I'd like to work on this one

**Comment by shaneutt** [comment]
You got it @arkodg, thank you for volunteering! I would definitely recommend the "What and why, but not the _how_ yet" approach to starting an initial GEP here so we can get aligned on what the problem is and what our goals are. Let us know if there's any help you need, and how we can support you in this effort! :vulcan_salute: 

/assign @arkodg

**Comment by arkodg** [comment]
thanks for the reminder @shaneutt, still plan on working on it

**Comment by shaneutt** [comment]
Sounds good, let us know if there's ways we can help facilitate :vulcan_salute:

**Comment by youngnick** [comment]
/remove-lifecycle stale
/lifecycle frozen

Since the GEP is still in motion.

**Comment by howardjohn** [comment]
@arkodg are there conformance tests planned?

**Comment by arkodg** [comment]
thanks for the reminder, let me create a issue to track it

**Comment by shaneutt** [comment]
Doesn't seem like we've had motion on this recently?

/remove-lifecycle frozen

@arkodg just wanted to check in, did you have any thoughts or a summary update for this one?

**Comment by arkodg** [comment]
next steps here would be 
1. have some implementations, implement this
2. building out conformance tests
3. make sure tests pass for at least 3 implementations
4. move to standard

we're still on 1., afaik, no one has implemented this yet

**Comment by shaneutt** [comment]
This feature has been accepted for the `v1.4.0` release for a special iteration. Please see [this announcement](https://github.com/kubernetes-sigs/gateway-api/discussions/3824) for more details.

**Comment by shaneutt** [comment]
@arkodg checking in: how are things going? Are you blocked on anything or need any support to help move this forward?

**Comment by arkodg** [comment]
@shaneutt would appreciate a review on https://github.com/kubernetes-sigs/gateway-api/pull/3838

**Comment by shaneutt** [comment]
This issue is targeting updates to `Standard` in the the [v1.4.0 release-cycle](https://github.com/kubernetes-sigs/gateway-api/discussions/3969), so this is just a reminder that we're looking to do code-freeze on August 26th, which is **two weeks from now**. @arkodg can you please provide an update on where we're at, and list all the outstanding issues and tasks?

**Comment by shaneutt** [comment]
@arkodg code freeze is tomorrow, checking in again?

**Comment by kl52752** [comment]
@shaneutt Client Certificate validation in v1.4.0 updates are targeting `Experimental` channel. All changes for GEP-91 were addressed in https://github.com/kubernetes-sigs/gateway-api/pull/3960#event-19239107393
@arkodg to confirm

**Comment by rouke-broersma** [comment]
I have been reading up on the different related GEPs and discussions around this and I feel like the persona that needs this capability has not been fully considered in all the discussions. The issue originally started with "As an application developer" however a lot of the discussion has been around how this would be implemented in the gateway config.

Application developers would not normally be able to modify the gateway listeners directly, this would normally (according to the responsibility model) fall under cluster operators. But client certificate authentication can be a very application developer requirement driven action. Some routes may require certificates whereas others may not. All other authentication efforts (ext auth) would be able to target HTTPRoute I think (I hope). Client certificate authentication is at the end of the day just another form of authentication, but this form of authentication can only apply (in current discussions) at the listener level. This requires strict cooperation between cluster operators and application developers to implement this requirement.

Ingress-NGINX allows per-ingress client certificate authentication. As far as I can tell with the current proposed implementation this would require a gateway listener per distinct authentication requirement. Our application developers will need to coordinate with us for each of their client authentication requirements, instead of being able to configure this independently. While support in-spec for client certificate validation is definitely useful for ingress-nginx -> gateway api migration, placing this on the gateway listener level created a significant bottleneck on the cluster operators for assisting with this migration.

Has this been considered? I couldn't find it anywhere.

**Comment by kl52752** [comment]
hi @rouke-broersma thank you for your feedback. Our number one priority was to address the security issue and unfortunately we couldn't find a way to leave client certificate validation configuration on Listener and eliminate the issue with connection coalescing. The best that we could do is to allow for Per port override. 

"As far as I can tell with the current proposed implementation this would require a gateway listener per distinct authentication requirement."
That's not totally true: All Listeners in the Gateway MAY share the same configuration but it can be overridden on per port basis. All Listeners with the same port must share the same client certificate configurations. 

The discussion for the issue was in the [doc]( 
https://docs.google.com/document/d/1qKIUyJsXI206dsubypEPfnxNLsHte4vtdjqZvAhMEDQ/edit?tab=t.0#heading=h.qbppdbrj8t36)

**Comment by rouke-broersma** [comment]
@kl52752 my concern is not whether it's at the gateway or listener level, but rather that it's on the gateway cr at all. This imo does not match with the expected user of this feature (application developer). 

Although I understand if there were practical or technical limitations why this must be on the gateway cr, I am missing the user story from the application developer perspective and the friction this limitation creates by forcing the cooperation between cluster operator and application developer.

Basically I'm asking was it considered that the most likely user role for this feature is the application developer role but that the cluster operator role is the one fully responsible for configuration?

**Comment by kl52752** [comment]
Oh I see. Looking at the [definition](https://gateway-api.sigs.k8s.io/concepts/roles-and-personas/#key-roles-and-personas) of Personas in gateway API:

> Chihiro (they/them) is a cluster operator, responsible for managing clusters to ensure that they meet the needs of their several users. Chihiro will typically be concerned with policies, network access, application permissions, etc. Again, they are beholden to no single user of any cluster; rather, they need to make sure that the clusters serve all users as needed.

I would put client certificate validation into `network access` and `application permissions`. It is sth more than allowing access to some `Application` this is authenticating clients for accessing routing resources sharing same Gateway.

**Comment by rouke-broersma** [comment]
I disagree. While cluster operators may enforce cluster wide policies like client certificate authentication, often teams have their own specific authentication requirements for certain routes independent of infrastructure level policies.

In this model the cluster operator has to configure new listeners for each ca the upstream team requires on their routes.

I'm also unsure if path based client certificate authentication is possible in the current proposal since the path is in the route but the client certificate is required on the whole listener? I am pretty sure this is also possible in ingress nginx.

Or, should this be solved with external auth instead?

**Comment by mahrton** [comment]
The original GEP is still open and seems to be connected to this ticket while GEP-91 is already mentioned in release 1.5.0. This ticket in turn points to release 1.4.0. Should this be cleaned up?

Edit: 
I would also like to mention that under https://gateway-api.sigs.k8s.io/geps/gep-91/#existing-support-in-implementations NGINX Gateway Fabric point at https://docs.nginx.com/nginx-ingress-controller/configuration/policy-resource/#ingressmtls which is Nginx Ingress documentation.

**Comment by Preisschild** [comment]
If this were to be implemented under the `ListenerTLSConfig` struct, it would also be usable in `ListenerSet`s, which means that application developers (not cluster admins) could validate client certificates with the CA secret in their own namespace.
